scio-core/src/main/scala/com/spotify/scio/values/BloomFilter.scala

```diff
@@ -0,0 +1,202 @@
+   *
+   * @return true if the element may be present, false if it is definitely not present.
+   */
+  def mayBeContains(t: T): Boolean = internal.mightContain(t)
+
+  // Java Serialization
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.writeObject(funnel)
+    internal.writeTo(out)
+  }
+
+  private def readObject(in: ObjectInputStream): Unit = {
+    val funnel = in.readObject().asInstanceOf[Funnel[T]]
+    val internal = gBloomFilter.readFrom(in, funnel)
+    BloomFilter.setField("funnel", funnel)
+    BloomFilter.setField("internal", internal)
+  }
+
+  /**
+   * Add an element to this [[BloomFilter]]. It creates a copy of the underlying
+   * structure.
+   *
+   * For creating BloomFilters from large collections, use [[BloomFilter#apply(Iterable)]]
+   * instead.
+   */
+  @experimental
+  def put(t: T): BloomFilter[T] = {
+    val copyOfInternal = internal.copy()
+    copyOfInternal.put(t)
+    this.copy(internal = copyOfInternal)
+  }
+
+  /**
+   * Add an Itearable of elements to this [[BloomFilter]]. It creates a copy of the underlying
+   * structure.
+   *
+   * For creating BloomFilters from large collections, use [[BloomFilter#apply(Iterable)]]
+   * instead.
+   */
+  @experimental
+  def putAll(elements: Iterable[T]): BloomFilter[T] = {
+    val copyOfInternal = internal.copy()
+    val it = elements.iterator
+    while (it.hasNext) {
+      copyOfInternal.put(it.next())
+    }
+    this.copy(internal = copyOfInternal)
+  }
+}
+
+object BloomFilter extends ApproxFilterCompanion {
+
+  /**
+   * Constructor for [[BloomFilter]]
+   *
+   * @param iterable An iterable of elements to be stored in the BloomFilter
+   * @param fpProb allowed false positive probability
+   * @param f [[Funnel]] for the type [[T]]
+   * @return A [[BloomFilter]] for the given [[Iterable]] of elements.
+   */
+  def apply[T](
+    iterable: Iterable[T],
+    fpProb: Double
+  )(
+    implicit f: Funnel[T]
+  ): BloomFilter[T] = {
+    val numElements = iterable.size
+    val settings = BloomFilter.optimalBFSettings(numElements, fpProb)
+    require(
+      settings.numBFs == 1,
+      s"BloomFilter overflow: $numElements elements found, max allowed: ${settings.capacity}"
+    )
+
+    val bf = gBloomFilter.create[T](f, numElements, fpProb)
+    val it = iterable.iterator
+    while (it.hasNext) {
+      bf.put(it.next())
+    }
+    BloomFilter(bf, f)
+  }
+
+  /**
+   * Constructor for an empty [[BloomFilter]]
+   *
+   * An empty BloomFilter is particularly useful when used with
+   * [[SCollection.asSetSingletonSideInput(defaultValue)]]
+   *
+   * @param numElements The number of elements that we might later want to insert using
+   *                    [[BloomFilter.put]] to create a new [[BloomFilter]]
+   * @param fpProb Expected false positive probability of the resulting [[BloomFilter]]
+   *               Default 0.01 (1 %)
+   */
+  def empty[T: Funnel](numElements: Int, fpProb: Double = 0.01): BloomFilter[T] = {
```

Comment:
If we disallow mutation (`put/putAll`) by default, this can be a `val` and we don't need any args?

Reply:
I am not sure I understand this.
So the idea here is that you can have an empty BF and then add an `Iterable` using `putAll` / `++`.

A `val` will mean a wrapper over a `Funnel`. we can't add elements without specifying a `fpProb`. How do you see this being used?

Reply:
The idea is to not expose `putAll`/`++`, and hence the ability to build a mutable empty BF.
In that case the `empty` instance should only be useful as a side input default in case the original collection is empty. Yeah so maybe `val` wouldn't work because of `T` but it might look this:

```
def empty[T] = new BloomFilter {
  override def maybeContains(v: T): Boolean = false
}
```